hazelcast/src/main/java/com/hazelcast/client/impl/spi/ProxyManager.java

```diff
@@ -330,13 +388,58 @@ private void initialize(ClientProxy clientProxy) throws Exception {
     }
 
     public Collection<? extends DistributedObject> getDistributedObjects() {
+        getAndProcessDistributedObjects();
+        return getLocalDistributedObjects();
+    }
+
+    private void getAndProcessDistributedObjects() {
+        ClientMessage request = ClientGetDistributedObjectsCodec.encodeRequest();
+        distributedObjectsWriteLock.lock();
+        try {
+            ClientInvocationFuture future = new ClientInvocation(client, request, client.getName()).invoke();
+            ClientMessage response = future.get();
+            processDistributedObjectInfos(response);
```

Comment:
If there are a lot of distributed objects in cluster, then this means number of objects times invocations to the cluster which may cause some delay and we are blocking all the proxy get operations, e.g. client.getMap, and this is not very desired, and that may be a hot path.